### ISSUE TITLE
📘 DOC: Better "Collections" examples and explanations, especially for those coming from other SSGs

### ISSUE THREAD

**Issue by brycewray** [snowpackjs/astro#973]
The "Collections" documentation, particularly https://docs.astro.build/core-concepts/collections#example-pagination, needs a lot of help if you're to attract anyone who's currently using another static site generator, especially one like Eleventy or Hugo that makes such activity much easier and doesn't require such deep knowledge of not only routing in general (as does Next.js, for example) but also how Astro does it in particular. For just a few examples of things I, as a definite non-expert where routing is concerned, have found to be a show-stopper thus far (especially since I also found little or no help about them in the [examples](https://github.com/snowpackjs/astro/tree/main/examples)):

(1.) **Globs?**

```js
const allPosts = Astro.fetchContent('../posts/*.md') // fetch local posts...
```
Would `('../posts/**/*.md')` be supported? If not, how would one fetch from deeper down? Many users have blogs with considerably deeper routes and, if migrating posts from another SSG to Astro, would rather set Astro to keep URLs the same rather than having to redirect willy-nilly. In my current site's case, the typical path for a post is `/posts/yyyy/mm/this-is-the-post-name`, and I don't think this is terribly uncommon. Also, anyone who wants to migrate content from a long-time WordPress site likely is also going to need similarly deep (or deeper) routing, so this kind of detail is vital.

In a similar vein...

(2.) **Wildcards for extensions?**

Would `('../posts/**/*.*')` be supported for those cases where one had a mix of straight Markdown files and Astro files? If you're bringing in years worth of Markdown files, it's likely you'll keep most of them that way except where certain content would benefit from the easier insertion of components through *.astro files (while, of course, keeping the same pre-extension file name and thus the same URL). Yes, I know there's talk of bringing back previously components-in-Markdown but I'm talking for now about only Astro's current capabilities.

(3.) **Non-paginated partial lists of collections?**

This refers to, say, listing only your three to five most recent posts on a page, such as the home page. You're not paginating; you want only a short list of the latest stuff. So my earlier questions and wishes for more details would apply here, too.

---

In short: while I realize this is 'waaay early in Astro's history and the team has all it can do to keep up with bugs and RFCs, I think it's also vital that questions like these be considered. Astro is riding a wave of good will right now as the SSG community thinks about its enormous potential, but its docs need to be friendlier where certain bread-and-butter SSG issues — like collections — come into play, and the questions I've asked here cover a few of those issues. The current collections documentation seems to assume too much about what the user knows. For those of us who lack the ability to figure out from Astro's source code how its routing works, we must rely on the docs. Thus, show-stoppers like this are particularly disappointing and, I hope, won't be left in the way for too long. Thanks in advance for any consideration this may receive.

**Comment by jasikpark** [comment]
I feel like it's probably also useful to add to the docs that you don't need to even acess the collections api if you don't want to: the docs website just adds a "layout: 'MainLayout.astro'" to each markdown file and that works fine

**Comment by rebelchris** [comment]
Perhaps we could use my article on this point to make it more clear?

https://daily-dev-tips.com/posts/how-to-render-individual-pages-from-an-external-collection/

**Comment by rebelchris** [comment]
The basic answer yes: 

```js
const allPosts = Astro.fetchContent('../posts/*.md')
```

Could you give me some more information on a use-case you would like to see covered, so I can try and write a bit more about that local file usecase?
As it is blogs for instance I think the current examples demonstrate it well, but maybe one should write out what happens?

**Comment by brycewray** [comment]
Appreciate the offer! (I'm just a user, not a contributor.)

Truthfully, I can't get even the current examples to work — my builds crash, even when I follow the [applicable example](https://docs.astro.build/core-concepts/collections#example-pagination) to the byte and give it the simplest possible structure over which to iterate (e.g., `/src/pages/posts/test.md` and/or `/src/pages/posts/test.astro`, much less any deeper levels) — which is why I asked my original question above. So I don't know that I can improve on what I said at the above. I find these are my show-stoppers right now.

In short, I want it explained as simply as the [Eleventy docs about collections](https://www.11ty.dev/docs/collections/). I admit that I'm spoiled. :) However, I suspect other potential users are in the same boat.

Sorry if that's not a helpful answer.

**Comment by rebelchris** [comment]
Hi Bryce, totally helpful answer! 
Let me see if I can explain this in more detail in the docs, I'm actually migrating from Eleventy to Astro, so pretty sure I'll be able to explain it in more detail.

**Comment by FredKSchott** [comment]
Thanks @rebelchris!

Re: issue 3. in your list above, we try to answer this in the current docs page:

> **Use a Collection when you need to generate multiple pages from a single template.** If you just want to generate a single page — like a long list linking to every post on your blog — then you can just use a normal page.

I agree that this could be a bit more clear though, especially what we mean by "use a normal page". But the tldr is that you wouldn't use a collection for that.

**Comment by jasikpark** [comment]
Can you try moving $posts.astro into the posts/ folder? maybe that would work??

**Comment by brycewray** [comment]
@jasikpark Nope, that didn't either. When I try going to `https://localhost:3000/posts/`, at least it doesn't crash but it simply `Not Found`. This is true regardless of whether I use `Astro.fetchContent('*.md')` (i.e., since it's now in the same folder) or the original `Astro.fetchContent('../posts/*.md')` Changes now pushed to the repo if you want to see.

**Comment by brycewray** [comment]
@jasikpark Now on 0.18.9. Looks like it works in build (https://astro-site.vercel.app/posts) but in dev mode I still get the `no matching path found` stuff I mentioned earlier. Just FYI.

**Comment by brycewray** [comment]
Repeating a question I [asked on Discord](https://discord.com/channels/830184174198718474/845451724738265138/873254937096761385), since the issue it describes falls somewhat into this discussion...

---

Are ternary operators (or even just conditionals, period) allowed within maps? For example, the following will build properly:
```js
      {posts.data.map((post) => (
        <h1><a href={post.url}>{post.title}</a></h1>
      ))}
```

...but the following...

```js
      {posts.data.map((post) => (
        <h1><a href={post.url}>{post.title}</a></h1>
        {post.subtitle
          ? `<p>${post.subtitle}</p>`
          : `<p>No subtitle here</p>`
        }
      ))}
```

...causes the build to break with an `UnhandledPromiseRejectionWarning` error message.

(Can't test in dev mode because of what I noted above.)

**Comment by FredKSchott** [comment]
I'm hopeful that the RFC for true file-based routing will be accepted on our RFC call next Tuesday, which will replace the Collections API entirely in v0.19.0 or v0.20.0, depending on when it lands. A lot of the early feedback has specifically called out that it feels easier to use and more familiar than our Collections API (it's a 1:1 match of how Next.js and SvelteKit do routing).

You can check it out here: https://github.com/snowpackjs/astro/pull/1010 
The docs are updated if you follow the preview URL in the first comment.
There's even an `npm install` instruction to help you try it today before it is released to everyone.

Hopefully this helps address the issues you outlined in your post! Really appreciate you taking the time to share all of this, I even referenced this post in the original RFC as evidence of need to keep improving Collections API :)

Closing this issue since these problems are addressed by #1010.

**Comment by brycewray** [comment]
Looking very much forward to seeing the new routing fleshed out in the [examples](https://docs.astro.build/getting-started)!
